addons/mass_mailing/models/mailing.py

```diff
@@ -1306,36 +1315,117 @@ def _unsubscribe_token(self, res_id, email):
+        """Store an image in an attachement and returns an url"""
+        attachment = self.env['ir.attachment'].create({
+            'datas': b64image,
+            'name': "cropped_image_mailing_{}".format(self.id),
+            'type': 'binary',})
+
+        attachment.generate_access_token()
 
-            attachment.generate_access_token()
+        return '/web/image/%s?access_token=%s' % (
+            attachment.id, attachment.access_token)
 
-            return '/web/image/%s?access_token=%s' % (
-                attachment.id, attachment.access_token)
+    def _get_image_by_url(self, url, session):
+        maxsize = int(tools.config.get("import_image_maxbytes", DEFAULT_IMAGE_MAXBYTES))
+        _logger.debug("Trying to import image from URL: %s", url)
+        try:
+            response = session.get(url, timeout=int(tools.config.get("import_image_timeout", DEFAULT_IMAGE_TIMEOUT)))
+            response.raise_for_status()
+
+            if response.headers.get('Content-Length') and int(response.headers['Content-Length']) > maxsize:
+                raise ImportValidationError(
+                    _("File size exceeds configured maximum (%s bytes)", maxsize)
+                )
+
+            content = bytearray()
+            for chunk in response.iter_content(DEFAULT_IMAGE_CHUNK_SIZE):
+                content += chunk
+                if len(content) > maxsize:
+                    raise ImportValidationError(
+                        _("File size exceeds configured maximum (%s bytes)", maxsize)
+                    )
+
+            image = Image.open(io.BytesIO(content))
+            w, h = image.size
+            if w * h > 42e6:  # Nokia Lumia 1020 photo resolution
+                raise ImportValidationError(
+                    _("Image size excessive, imported images must be smaller than 42 million pixel")
+                )
 
-        modified = False
-        root = lxml.html.fromstring(body_html)
+            return content
+        except Exception as e:
+            _logger.exception(e)
+            raise ImportValidationError(
+                _(
+                    "Could not retrieve URL: %(url)s: %(error)s",
+                    url=url, error=e
+                )
+            )
+
+    def _convert_inline_images_to_urls(self, root):
+        """
+        Find inline base64 encoded images, make an attachement out of
+        them and replace the inline image with an url to the attachement.
+        """
+        did_convert = False
         for node in root.iter('img'):
             match = image_re.match(node.attrib.get('src', ''))
             if match:
                 mime = match.group(1)  # unsed
                 image = match.group(2).encode()  # base64 image as bytes
 
-                node.attrib['src'] = _image_to_url(image)
-                modified = True
+                node.attrib['src'] = self._image_to_url(image)
+                did_convert = True
+        # handle the images in mso comments
```

Comment:
Would recommand building a list of images (and their related nodes) to create, then create all attachments, then updating the nodes. That way we are batch-enabled, which would be better from performance point of view. And also try to iterate once to spot nodes to update, then once to udpate (instead of iterating several times, doing both).

Reply:
Done here: https://github.com/odoo/odoo/pull/119325/commits/626b81c37fc235b5d6194905203cf50269a3544c